pkg/workloadmeta/collectors/internal/kubeapiserver/reflector_store.go

```diff
@@ -0,0 +1,211 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver
+// +build kubeapiserver
+
+package kubeapiserver
+
+import (
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/DataDog/datadog-agent/pkg/workloadmeta"
+)
+
+type reflectorStore struct {
+	wlmetaStore workloadmeta.Store
+
+	mu   sync.Mutex
+	seen map[string]workloadmeta.EntityID
+}
+
+func newReflectorStore(wlmetaStore workloadmeta.Store) cache.Store {
+	return &reflectorStore{
+		wlmetaStore: wlmetaStore,
+		seen:        make(map[string]workloadmeta.EntityID),
+	}
+}
+
+// Add notifies the workloadmeta store with  an EventTypeSet for the given
+// object.
+func (r *reflectorStore) Add(obj interface{}) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	pod := obj.(*corev1.Pod)
+	entity := parsePod(pod)
+
+	r.seen[string(pod.UID)] = entity.EntityID
+
+	r.wlmetaStore.Notify([]workloadmeta.CollectorEvent{
+		{
+			Type:   workloadmeta.EventTypeSet,
+			Source: collectorID,
+			Entity: entity,
+		},
+	})
+
+	return nil
+}
+
+// Update notifies the workloadmeta store with  an EventTypeSet for the given
+// object.
+func (r *reflectorStore) Update(obj interface{}) error {
+	return r.Add(obj)
+}
+
+// Delete notifies the workloadmeta store with  an EventTypeUnset for the given
+// object.
+func (r *reflectorStore) Delete(obj interface{}) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	pod := obj.(*corev1.Pod)
+
+	delete(r.seen, string(pod.UID))
+
+	r.wlmetaStore.Notify([]workloadmeta.CollectorEvent{
+		{
+			Type:   workloadmeta.EventTypeUnset,
+			Source: collectorID,
+			Entity: &workloadmeta.KubernetesPod{
+				EntityID: workloadmeta.EntityID{
+					Kind: workloadmeta.KindKubernetesPod,
+					ID:   string(pod.UID),
+				},
+			},
+		},
+	})
+
+	return nil
+}
+
+// Replace diffs the given list with the contents of the workloadmeta store
+// (through r.seen), and updates and deletes the necessary objects.
+func (r *reflectorStore) Replace(list []interface{}, _ string) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	var events []workloadmeta.CollectorEvent
+
+	seenNow := make(map[string]workloadmeta.EntityID)
+	seenBefore := r.seen
+
+	for _, obj := range list {
+		pod := obj.(*corev1.Pod)
+		podUID := string(pod.UID)
+		entity := parsePod(pod)
+
+		events = append(events, workloadmeta.CollectorEvent{
```

Comment:
Can this create duplicate events? Not sure if this can create "set" events for pods for which we already sent a "set" event.